hw_07.09.23/task2/phonebook.py

```diff
@@ -0,0 +1,138 @@
+
+if os.path.exists(datafile_path):
+    with open(datafile_path, "r") as file:
+        PHONEBOOK = json.load(file)
+
+
+def add_contact():
+    first_name = input("Enter first name: ")
+    last_name = input("Enter last name: ")
+    full_name = first_name + " " + last_name
+    telephone_number = input("Enter telephone number: ")
+    city_or_state = input("Enter city or state: ")
+
+    if telephone_number and full_name:
+        PHONEBOOK[telephone_number] = {
+            "first_name": first_name,
+            "last_name": last_name,
+            "full_name": full_name,
+            "address": city_or_state,
+        }
+
+        with open(datafile_path, "w") as file_obj:
+            json.dump(PHONEBOOK, file_obj, indent=4)
+            print(PHONEBOOK)
+
+    else:
+        print("Invalid input. Both name and telephone number are required.")
+
+
+def search_by_fullname(type_: str) -> dict:
+    return {number: contact for number, contact in PHONEBOOK.items() if type_.lower() in contact['full_name'].lower()}
+
+
+def search_by_phone(type_: str) -> dict:
+    return {number: contact for number, contact in PHONEBOOK.items() if type_.lower() in number.lower()}
+
+
+def search_by_city(type_: str) -> dict:
+    return {number: contact for number, contact in PHONEBOOK.items() if type_.lower() in contact['address'].lower()}
+
+
+def update_phonebook(phone):
+    first_name = input("Enter first name: ")
+    last_name = input("Enter last name: ")
+    full_name = first_name + " " + last_name
+    city_or_state = input("Enter city or state: ")
+
+    updated_phonebook = {}
+    if phone and full_name:
+        updated_phonebook[phone] = {
+            "first_name": first_name,
+            "last_name": last_name,
+            "full_name": full_name,
+            "address": city_or_state,
+        }
+        PHONEBOOK.update(updated_phonebook)
+
+        with open(datafile_path, "w") as file_obj:
+            json.dump(PHONEBOOK, file_obj, indent=4)
+            print(PHONEBOOK)
+
+
+def delete_phone():
+    number = input("Enter telephone number to delete: ")
+    if number in PHONEBOOK:
+        del PHONEBOOK[number]
+
+        with open(datafile_path, "w") as file_obj:
+            json.dump(PHONEBOOK, file_obj, indent=4)
+
+        print(f"Phone number {number} deleted.")
+    else:
+        print(f"Phone number {number} not found in the phonebook.")
+
+
+while True:
+
+    print("Please select options")
+    print("+-----------------+")
+    print("1: Add new entries")
+    print("2: Search by name, surname, full name")
+    print("3: Search by telephone number")
+    print("4: Search by city or state")
+    print("5: Delete a record for a given telephone number")
+    print("6: Update a record for a given telephone number")
+    print("7: Exit the program")
+    print("+------------------+")
+
+    user_chose = input("Chose number: ")
+    allowed_input = ["1", "2", "3", "4", "5", "6", "7"]
```

Comment:
Because you have same logic spread into two variables you can easily forget to update `allowed_input` after you add new `if` with a new option. I'd suggest the following code:
```
    action_by_choice = {
        "1": add_contact,
        "2": search_by_fullname,
        "3": search_by_phone,
        "4": search_by_city,
        "5": delete_phone,
        "6": update_phonebook,
        "7": exit,
    }

    if user_choice in allowed_input:
        action_by_choice[user_choice]()
        print("===========>")
    else:
        "Please enter correct number"
```
That's much-much shorter. For this to work, you need to update functions and put respective inputs inside each function.